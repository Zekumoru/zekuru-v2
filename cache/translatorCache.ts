import { AuthorizationError, Translator, TranslatorOptions } from 'deepl-node';
import { Collection } from 'discord.js';
import GuildKey, { IGuildKey } from '../models/GuildKey';
import { loadLanguages } from '../translation/languages';
import { errorDebug } from '../utils/logger';
import { decrypt, encrypt } from '../utils/crypt';

export const translatorOptions: TranslatorOptions = {
  appInfo: {
    appName: 'Zekuru-v2 Demo',
    appVersion: '1.0.0',
  },
  minTimeout: 500, // 500 ms
  maxRetries: 10,
};

const cacheTranslators = new Collection<string, Translator>();

const fetchOrCreateGuildKey = async (guildId: string, hashedApiKey: string) => {
  // fetch from db, if not exists, create new
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (guildKey) return [guildKey, false] as const;

  // save api key to db
  const newGuildKey = new GuildKey({
    id: guildId,
    key: hashedApiKey,
  });
  await newGuildKey.save();

  return [newGuildKey, true] as const;
};

const set = async (guildId: string, apiKey: string) => {
  // create translator client
  const translator = new Translator(apiKey, translatorOptions);

  // this should throw error if invalid api key
  await translator.getUsage();
  await loadLanguages(translator);

  // encrypt api key
  const encryptedApiKey = encrypt(apiKey);

  // update in db
  const [guildKey, isNewlyCreated] = await fetchOrCreateGuildKey(
    guildId,
    encryptedApiKey
  );

  if (!isNewlyCreated) {
    guildKey.overwrite({
      id: guildId,
      key: encryptedApiKey,
      createdAt: guildKey.createdAt,
    } as IGuildKey);
    await guildKey.save();
  }

  // update cache
  cacheTranslators.set(guildId, translator);
};

const get = async (guildId: string) => {
  // get from cache
  const translatorCache = cacheTranslators.get(guildId);
  if (translatorCache) return translatorCache;

  // if not exists in cache, fetch from api key from db
  // then create new translator and set to cache
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (!guildKey) return;

  try {
    const apiKey = decrypt(guildKey.key);
    const translator = new Translator(apiKey, translatorOptions);

    // load languages
    await loadLanguages(translator);

    // update cache
    cacheTranslators.set(guildId, translator);
    return translator;
  } catch (error) {
    // delete invalid api key from db and "sign out" user by removing it from cache
    if (error instanceof AuthorizationError) {
      await unset(guildId);
    }
    errorDebug(error);
  }
};

const unset = async (guildId: string) => {
  // check if it exists first
  if ((await get(guildId)) == null) return;

  // delete from db
  await GuildKey.deleteOne({ id: guildId });

  // remove from cache
  cacheTranslators.delete(guildId);
};

export default { set, unset, get };
