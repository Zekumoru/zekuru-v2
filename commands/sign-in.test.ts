import signIn from './sign-in';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import translatorCache from '../cache/translatorCache';
import { AuthorizationError } from 'deepl-node';

jest.mock('../cache/translatorCache.ts');

describe('/sign-in command', () => {
  const mockTranslatorCache = jest.mocked(translatorCache);

  afterEach(() => {
    jest.resetAllMocks();
  });

  it(`should notify user that this command is only for servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await signIn.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they are already signed in`, async () => {
    const interaction = createMockInteraction();
    mockTranslatorCache.get.mockResolvedValue({} as any);

    await signIn.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they didn't provide an API key`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue(null);

    await signIn.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they passed an invalid API key`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('invalid key');
    mockTranslatorCache.set.mockImplementation(async () => {
      throw new AuthorizationError('Invalid API key!');
    });

    await signIn.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if something went wrong`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('valid key');
    mockTranslatorCache.set.mockImplementation(async () => {
      throw new Error('Generic error!');
    });

    await signIn.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should sign in`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getString.mockReturnValue('valid key');

    await signIn.execute(interaction);

    expect(mockTranslatorCache.set).toHaveBeenCalled();
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });
});
