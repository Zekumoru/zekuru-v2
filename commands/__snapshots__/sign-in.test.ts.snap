// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/sign-in command should notify user if something went wrong 1`] = `
{
  "content": "Bot could not sign in. Something went wrong!",
}
`;

exports[`/sign-in command should notify user if they are already signed in 1`] = `
{
  "content": "You are already signed in. Please sign out using \`/sign-out\` command if you wish to sign in with another API key.",
}
`;

exports[`/sign-in command should notify user if they didn't provide an API key 1`] = `
{
  "content": "Please provide a Deepl API key.",
}
`;

exports[`/sign-in command should notify user if they passed an invalid API key 1`] = `
{
  "content": "Bot could not sign in. Invalid Deepl API key!",
}
`;

exports[`/sign-in command should notify user that this command is only for servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/sign-in command should sign in 1`] = `
{
  "content": "Bot has been successfully signed in. You can now translate!",
}
`;
